README.md
# Board-Games-Results

Saving your results for 7 Wonders games in a simple web based application

[![Build Status](https://travis-ci.org/GorlifSense/Board-Games-Results.svg?branch=master)](https://travis-ci.org/GorlifSense/Board-Games-Results)
[![codecov](https://codecov.io/gh/GorlifSense/Board-Games-Results/branch/master/graph/badge.svg)](https://codecov.io/gh/GorlifSense/Board-Games-Results)
[![Dependency Status](https://www.versioneye.com/user/projects/5797b61474848d002b4b9572/badge.svg?style=flat-square)](https://www.versioneye.com/user/projects/5797b61474848d002b4b9572)
[![NSP Status](https://nodesecurity.io/orgs/gorlifsense/projects/9e48a233-c94e-48bc-84a6-3cb12c3d1142/badge)](https://nodesecurity.io/orgs/gorlifsense/projects/9e48a233-c94e-48bc-84a6-3cb12c3d1142)

## Quickstart

Set `MONGO` environment variable to access your MongoDB.
Instead, default `mongodb://localhost/boardgamesresults` is used.

`npm install`  
`npm start`  

Here it is. You should have running it locally on `http://localhost:3001`

### Demo

You can view application in development mode launched
[here…](https://boardgamesresults.herokuapp.com)

## Development

[![StackShare](http://img.shields.io/badge/tech-stack-0690fa.svg?style=flat)](http://stackshare.io/Fleischers/board-games-results)

Project has guidelines to follow.
Be ready to check output of pre-commit hooks
to keep consistency throughout all files.

We have Continuous Deployment enabled and to avoid broken commits,
`master` branch will be frozen
and all code should be committed to feature branches
or `development`. We would like to follow here common practices
[More on Github Flow…](https://guides.github.com/introduction/flow/)

### Code Quality

[![bitHound Overall Score](https://www.bithound.io/github/GorlifSense/Board-Games-Results/badges/score.svg)](https://www.bithound.io/github/GorlifSense/Board-Games-Results)
[![Issue Count](https://codeclimate.com/github/GorlifSense/Board-Games-Results/badges/issue_count.svg)](https://codeclimate.com/github/GorlifSense/Board-Games-Results)

Use `.editorconfig` for your Code Editor to keep code style.  

Use `npm run beautify -- <filename> -r --type <js|html|css>`
for auto format code.
Replace `<filename>` with file to beautify
and pick `<js|html|css>` language type.
[Read more options…](https://www.npmjs.com/package/js-beautify)

Run `npm run eslint` to check common javascript mistakes.  

Run `npm run csslint` to check common css mistakes.  

Run `npm run remark-lint` to check markdown.
[Read more about rules…](https://github.com/wooorm/remark-lint/blob/master/doc/rules.md)

### API

View `spec/` directory for most recent RAML specification.
[HTML generated version available…](http://gorlifsense.com/Board-Games-Results/)

#### Basic routes

Current routes state are unsecured
and not all error messages included in response

| PATH                       | Description        |
| :-------------             | :-------------     |
| GET /tables                | List of tables     |
| POST /tables               | Create new table   |
| GET /tables/:tableId       | Find table by id   |
| PUT /tables/:tableId       | Edit table by id   |
| DELETE /tables/:tableId    | Delete table by id |

### Tests

`npm test` - to run tests  
Look into `open coverage/lcov-report/index.html` to check coverage report

## Contributing

1. Fork and clone it
2. Install dependencies: `npm install`
3. Create a feature branch: `git checkout -b new-feature`
4. Commit changes: `git commit -am 'Added a feature'`
5. Run unit tests: `npm test`
6. Run static code analysis `npm run eslint`
7. Push to the remote branch: `git push origin new-feature`
8. Create a new [Pull Request](https://github.com/GorlifSense/Board-Games-Results/pull/new/master)

## License

Code released under the [ISC license](https://github.com/GorlifSense/Board-Games-Results/blob/master/LICENSE).


CHANGELOG.md
Release notes for Board Games Results
====================================

Planned in next minor version
------------------

* Support Angular2 app and remove @deprecated
* Finish changed Table structure

v0.1.9
------------------

* Added Dockerfile and docker-compose

v0.1.8
------------------

* Enabled CORS
* Autofix ESLint
* Updated few devDependencies

v0.1.7
------------------

* Timely fix for Heroku.

v0.1.6
------------------

* Fixed ESLint errors
* Formatting replaced tabs with spaces
* Enabled LGTM integration
* Button "back" on create.html to Default instead of Danger
* Fix Codecov configuration

v0.1.5
------------------

* Remove unnecessary create.html template
* Add createdBy property to Table edit and creation
* Do not ignore tests in Code Climate
* Update test table generator to upcoming Table structure
* Updated RAML for upcoming Table structure

*Known Issues*:

* Found design problem in current model Test Suites

v0.1.4
------------------

* Prepare @deprecated methods for deletion
* Update remark-lint rules
* Heroku generated app.json added to support pipelines
* Fix 'resource' issue in RAML. Additional description to API
* VersionEye now works for teams and as integration

v0.1.3
------------------

* Revert visual fixes in /tables

v0.1.2
------------------

* Updated Eslint to v3
* Visual fixes by Andrew

v0.1.1
------------------

* Added lodash dependency to prevent application 503 error

v0.1.0
------------------

* Unsecured CRUD API with Koa v1:
  Create, Read, Update, Delete on Table collection

v0.0.4
------------------

* Placed greeting line
* Removed static HTML template test on main page.
* Remove unused index.html scripts
* Remove support for IE

v0.0.3
------------------

* Using ESLint v2
* Added Codecov config, Remarked lint, CSSLint, .nvmrc
* bitHound
* configured Pull Request hooks
* RAML docs
* [#11 Fix tables template](https://github.com/GorlifSense/Board-Games-Results/pull/11)

v0.0.2 and before
------------------

* Create Table
* View Tables


test/init.js
'use strict';

const app = require('../app');

describe('Dummy', () => {

  it('should try launch server and testd', () => {
    const mongoUrl = 'mongodb://localhost:27017/boardgamesresults_test';

    app('3001', process.env.MONGO || mongoUrl);

  });

});


test/tables.js
'use strict';

const Mongorito = require('mongorito');
const should = require('chai').should();

const _ = require('lodash');

const tableGenerator = require('./fixtures/tables');
const tables = require('../controllers/tables');

require('mocha-generators').install();
require('../helpers/logger');
// TODO remove Winston
const winston = require('winston');

describe('Tables controller', () => {

  before(() => {
    const mongoDBUrl = 'mongodb://localhost/boardgamesresults';

    Mongorito.connect(process.env.MONGO || mongoDBUrl);
    tables.assert = (params, statusCode, stringLine) => {
      // should.exist(params);
    };

  });

  it('should createTable', function *test1() {

    const generatedTable = tableGenerator();

    // generatedTable._id = '5771ba42209301cc089f43dd';
    tables.body = generatedTable;
    winston.debug('tables == ', tables);

    yield tables.add();
    winston.silly('tables == ', tables);

    should.exist(tables.body.data);
    tables.body.success.should.be.equal(true);
    tables.body.status.should.be.equal(200);
    tables.body.data.attributes.createdBy.should.be.equal('anonymous');
    tables.status.should.be.equal(200);

  });

  it.skip('should edit table and find it', function *test4b() {
    tables.params = {
      tableId: '5771ba42209301cc089f43dd'
    };
    tables.body.createdBy = '1377ba42209301cc089f43d9';
    yield tables.edit();

    winston.debug('tables == ', tables);

    should.exist(tables.body.message);
    // TODO FINISH TEST
    // tables.body.status.should.be.equal(200);
    // tables.status.should.be.equal(200);
    // tables.body.data.attributes
    //   .createdBy.should.be.equal('1377ba42209301cc089f43d9');

  });

  it('should get table and not found it', function *test2() {

    tables.params = {
      tableId: '5771ba42209301cc089f43d9'
    };
    yield tables.get();

    should.exist(tables.body.message);
    tables.body.success.should.be.equal(false);
    tables.body.status.should.be.equal(404);
    tables.status.should.be.equal(404);
  });

  it('should delete table and not found it', function *test3() {

    tables.params = {
      tableId: '5771ba42209301cc089f43d9'
    };
    yield tables.remove();

    should.exist(tables.body.message);
    tables.body.success.should.be.equal(false);
    tables.body.status.should.be.equal(404);
    tables.status.should.be.equal(404);
  });

  it('should edit table but not found it', function *test4() {

    tables.params = {
      tableId: '5771ba42209301cc089f43d9'
    };

    yield tables.edit();

    should.exist(tables.body.message);
    tables.body.success.should.be.equal(false);
    tables.body.status.should.be.equal(404);
    tables.status.should.be.equal(404);
  });

  it('should list all tables', function *test5() {
    yield tables.list();
    tables.body.should.be.an('array');
    winston.debug('tables count == ', tables.body.length);

  });

  it.skip('should reduce all results to one number', function *test6() {

    yield tables.getTables();
    const firstPlayer = tables.data[0].attributes.game.players[0];

    firstPlayer.situation.should.be.an('number');
    winston.debug('player 1 situation == ', firstPlayer.situation);

  });

});


test/fixtures/tables.js
'use strict';

const chance = require('chance')();


class Player {
  constructor() {
    this.fullname = chance.first();
    this.city = {
      name: chance.pickone([
        'Gizah',
        'Babylon',
        'Ephesos',
        'Olympia',
        'Halicarnassus',
        'Rhodes',
        'Alexandria'
      ]),
      side: chance.pickone(['A', 'B'])
    };
    this.situation = {
      'military': {
        'score': chance.integer()
      },
      'gold': {
        'coins': chance.integer(),
        'score': chance.integer()
      },
      'wonder': {
        'score': chance.integer()
      },
      'culture': {
        'score': chance.integer()
      },
      'trade': {
        'score': chance.integer()
      },
      'guild': {
        'score': chance.integer()
      },
      'science': {
        'cog': chance.integer(),
        'tablet': chance.integer(),
        'compass': chance.integer(),
        'score': chance.integer()
      }
    };
  }
}

/**
 * Generate array with random players
 * @param  {number} count - integer of players on the table
 * @return {array}        - Players
 */
function generatePlayers(count) {
  const players = [];
  const TIMES = 1;

  while (count) {
    // Prepare array of Players around the table
    players.push(new Player());
    count -= TIMES;
  }
  return players;
}

/**
 * Expose table fixture
 * @return {object} - Table generated with players
 */
module.exports = function tableFixture() {
  return {
    description: chance.paragraph(),
    game: {
      title: '7 Wonders',
      players: generatePlayers(chance.integer({min: 3, max: 7}))
    }
  };
};


app.js
'use strict';

// Global dependencies
const app = require('koa')();
const logger = require('koa-logger');
const route = require('koa-route');
const Router = require('koa-router');
const json = require('koa-json');
const favicon = require('koa-favicon');
const serve = require('koa-static');
const cors = require('koa-cors');
const koaBetterBody = require('koa-better-body');
const compress = require('koa-compress');
const Mongorito = require('mongorito');

// Project dependencies
const Tables = require('./controllers/tables');
const router = require('./router');

// Internal variables
// @deprecated
const getTables = Tables.getTables;
const postTables = Tables.postTables;

// logger
// TODO use external log provider https://trello.com/c/IslYoZ2u
app.use(logger());

// use compression
app.use(compress());

// CORS
app.use(cors());

/**
 * response with static
 * Serving files will be @deprecated after Angular2 release
 */
app.use(favicon('public/favicon.ico'));
app.use(serve('public', {
  maxage: 10000
}));


/**
 * Generate and render pages with swig engine
 * NOTE Will be @deprecated after Angular2 release
 */
app.use(route.get('/tables', getTables));
app.use(route.post('/tables', postTables));

// API Routes
const api = new Router({
  prefix: '/api'
});
const Boom = require('boom');

// TODO remove pretty print
api.use(json());

router.use(koaBetterBody());

// TODO change this to more general
api.get('/', function *apiRoot() {
  this.body = {
    success: true,
    status: 200,
    message: 'Read the docs of API',
    link: 'http://gorlifsense.com/Board-Games-Results'
  };
});

api.use('/v:version', router.routes());

app.use(api.routes());
app.use(api.allowedMethods({
  'throw': false,
  notImplemented: () =>
    new Boom.notImplemented('that method is not implemented'),
  methodNotAllowed: () =>
    new Boom.methodNotAllowed('that method is not allowed')
}));

/**
 * Connect to Mongo and start app listening to port
 * @param  {string} PORT  - port of web server
 * @param  {string} MONGO - host of mongodb instance 'mongodb://host/db'
 */
module.exports = function startServer(PORT, MONGO) {

  Mongorito.connect(MONGO);
  app.listen(PORT);

};


public/js/main.js
'use strict';
// NOTE this will be @deprecated after Angular 2 based SPA released

// TODO Fix ESLint warnings
// https://trello.com/c/ex9mE2WS/57-fix-eslint-issues-in-main-js
document.addEventListener('DOMContentLoaded', () => {
  const getE = function (query, parent) {
    const selector = query.charAt(0);
    const name = query.slice(1);
    const element = parent || document;

    if (selector === '#') {
      return document.getElementById(name);
    } else if (selector === '.') {
      return element.getElementsByClassName(name);
    }
    return element.getElementsByTagName(query);

  };
  const playerForm = getE('.player-none')[0];
  const addPlayer = getE('#add-player');
  const submitButton = getE('#confirm-table');
  let playerNumber = 1;
  // Player add function

  addPlayer.addEventListener('click', () => {
    if (playerNumber < 7) {
      playerNumber++;
      const players = getE('.player');
      const newPlayer = playerForm.cloneNode(true);
      const removeButton = document.createElement('button');

      newPlayer.className = 'player';

      removeButton.className = 'remove-player btn btn-danger';
      removeButton.innerHTML = '-';
      removeButton.addEventListener('click', () => {
        newPlayer.style.animationName = 'playerFade';
        newPlayer.style.animationDuration = '.5s';
        const x = setTimeout(() => {
          playerNumber--;
          players[0].parentNode.removeChild(newPlayer);
        }, 500);
      });
      newPlayer.appendChild(removeButton);

      const controls = getE('.controls')[0];

      players[0].parentNode.insertBefore(newPlayer, controls);
    }

  });

  // Form submit function
  function collectForm() {
    const players = getE('.player');
    const object = Object.create(null);
    let error = null;
    const errorElement = getE('#error');

    object.name = getE('#js-table-name').value;
    object.game = {};

    function collect(nodeCollection, object) {
      Array.prototype.forEach.call(nodeCollection, (node) => {
        if (node.name && node.value && node.value !== '' && node.name !==
          'name') {
          object[node.name] = node.value;
        } else if (node.value === '') {
          if (!error) {
            error = 'Fill all required fields first.';
          }
          node.className += ' invalid';
          if (node.className.indexOf('nodeInvalidRemoveFunction') < 0) {
            node.className += ' nodeInvalidRemoveFunction';
            node.addEventListener('click', function () {
              this.className = this.className.replace(/\s?invalid/, '');
              errorElement.className = '';
            });
          }
        }
      });
    }

    // Description
    const description = document.getElementsByTagName('textarea')[0];

    object.description = description.value;

    // Players and city
    object.game.players = [];
    Array.prototype.forEach.call(players, (player, index) => {

      const currentPlayer = {};
      const name = player.getElementsByClassName('player-name')[0];

      currentPlayer.name = name.value;

      let cityObject = player.getElementsByTagName('select')[0];

      cityObject = cityObject.value.match(/(\w+)\((\w)\)/);

      currentPlayer.city = {
        'name': cityObject[1],
        'side': cityObject[2]
      };


      const inputs = player.getElementsByTagName('input');

      currentPlayer.situation = {};
      collect(inputs, currentPlayer.situation);

      object.game.players.push(currentPlayer);
    });
    if (!error) {
      // NOTE: in version v0.0.1 simple HTML POST will do fine
      $.post('/tables', object);
      window.location.href = '/tables'; /* We can do just that after callback and popup alert about table creation*/
    } else {
      errorElement.className = 'active';
      errorElement.innerHTML = error;
    }
  }
  // Remove placeholders to values
  const input = getE('input');

  if (input.length > 0) {
    Array.prototype.forEach.call(input, (item, index) => {
      item.__initialPlaceholder = item.placeholder || item.value;
      if (!item.__initialPlaceholder) {
        return;
      }
      item.addEventListener('focus', function () {
        if (this.value === this.__initialPlaceholder) {
          this.value = '';
        }
      });
      item.addEventListener('focusout', function () {
        if (this.value === '') {
          this.value = this.__initialPlaceholder;
        }
      });
    });
  }
  submitButton.addEventListener('click', collectForm);
});


controllers/tables/index.js
'use strict';

const parse = require('co-body');
const winston = require('winston');
const _ = require('lodash'); // eslint-disable-line id-length

const render = require('../../render');
const Table = require('./model');

const ok = 200;
const badRequest = 400;
const notFound = 404;

const CommonResponse = {
  success: true,
  status: ok
};


exports.list = function *listTables() {
  this.status = ok;
  this.body = yield Table.all();
};

exports.add = function *createTable() {
  const body = this.body;

  this.assert(body, badRequest, 'Body is empty');
  this.assert(body.game, badRequest, 'Body.game is empty');
  const table = new Table({
    _id: body._id, // HACK to perform tests over edit tables
    description: body.description,
    game: body.game,
    createdBy: 'anonymous' // TODO Change after User logic is implemented
  });
  const response = _.cloneDeep(CommonResponse);

  response.data = yield table.save();
  this.status = ok;
  winston.silly(response);
  this.body = response;
};

exports.get = function *getOne() {
  const params = this.params;

  this.assert(params, badRequest, 'Params are empty');
  const table = yield Table.where('_id', params.tableId).findOne();

  const response = _.cloneDeep(CommonResponse);

  if (table) {
    response.data = table;
  } else {
    response.message = 'Table not found';
    this.status = response.status = notFound;
    response.success = false;
  }
  winston.silly(response);
  this.body = response;
};

exports.edit = function *editTable() {
  const body = this.body;
  const params = this.params;

  winston.info('Body and params', body, params);

  this.assert(body, badRequest, 'Body is empty');
  this.assert(params, badRequest, 'Params are empty');
  const table = yield Table.where('_id', params.tableId).findOne();

  winston.info('table is', table);
  const response = _.cloneDeep(CommonResponse);

  if (table) {
    if (body.description) {
      table.set('description', body.description);
    }
    if (body.game) {
      table.set('game', body.game);
    }
    if (body.createdBy) {
      table.set('createdBy', body.createdBy);
    }

    response.data = yield table.save();
  } else {
    this.status = response.status = notFound;
    response.success = false;
    response.message = 'Table not found';
  }

  winston.silly(response);
  this.body = response;
};

exports.remove = function *removeTable() {
  const params = this.params;

  this.assert(params, badRequest, 'Params are empty');
  winston.debug(params);

  const table = yield Table.where('_id', params.tableId).findOne();

  const response = _.cloneDeep(CommonResponse);

  if (table) {
    table.set('deleted', true);
    const removal = yield table.remove();

    response.data = removal;
  } else if (_.isEmpty(table)) {
    this.status = response.status = notFound;
    response.success = false;
    response.message = 'Table not found, it could be removed before';
  }

  winston.silly(response);
  this.body = response;
};

/**
 * Get Tables
 * @deprecated
*/
function *getTables() {

  const tables = yield Table.all();

  // FIXME 500 Error.
  // TypeError: Cannot read property 'players' of undefined
  //    at tables.forEach
  //    (/Users/fcl/src/Board-Games-Results/controllers/tables/index.js:127:15)
  //
  // tables.forEach((table) => {
  //
  //   table.attributes.game.players.forEach((player) => {
  //
  //     const keys = Object.keys(player.situation);
  //
  //     if (keys) {
  //
  //       player.situation = keys.reduce((aVar, bVar) => {
  //         if (typeof aVar !== 'number') {
  //           aVar = Number(player.situation[aVar]);
  //         }
  //         return aVar + Number(player.situation[bVar]);
  //       });
  //     }
  //
  //   });
  // });

  winston.silly(tables);

  this.data = tables;
  this.body = yield render('tables', {tables});

}

/**
 * Post Table
 * @deprecated
 */
function *postTables() {

  const table = new Table(yield parse(this));

  winston.silly(table);
  yield table.save();
  this.redirect('/tables');

}

exports.getTables = getTables;
exports.postTables = postTables;
